### PR TITLE
*: update stale message check.

### PR DESCRIPTION
**Pull request by siddontang**
Still for #417 

@ngaut @BusyJay

### CODE REVIEW

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -245,19 +245,22 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
-                    return Ok(());
-                }
+        // If we receive a message which peer is not in current region and
+        // epoch is stale, we can ignore it.
+        if let Some(peer) = self.region_peers.get(&region_id) {
+            let from_epoch = msg.get_region_epoch();
+            let conf_ver = peer.storage.rl().region.get_region_epoch().get_conf_ver();
+            let not_exist = util::find_peer(&peer.storage.rl().region, from.get_store_id())
+                                .is_none();
+            // Is it necessary to check version too?
```
Comment (ngaut) [inline src/raftstore/store/store.rs]: We should check region version to filter unnecessary message.

File: src/raftstore/store/store.rs [hunk 2]
```diff
@@ -245,19 +245,22 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+            // Should we need to check RequestVote even peer in this region too?
+            if from_epoch.get_conf_ver() < conf_ver && not_exist {
+                warn!("RequestVote with a stale epoch {:?}, ignore it", from_epoch);
```
Comment (ngaut) [inline src/raftstore/store/store.rs]: The message maybe  raft response message, may be not RequestVote.

Comment (siddontang) [inline src/raftstore/store/store.rs]: If the peer is not in region peer lists, and conf version is stale, I think we can ignore all messages.

Comment (siddontang) [inline src/raftstore/store/store.rs]: btw, in raft, if the peer is not found and the message is response, an error returns.
```
// Step advances the state machine using the given message.
    pub fn step(&mut self, m: Message) -> Result<()> {
        // ignore unexpected local messages receiving over network
        if is_local_msg(&m) {
            return Err(Error::StepLocalMsg);
        }
        if self.raft.prs.contains_key(&m.get_from()) || !is_response_msg(&m) {
            return self.raft.step(m);
        }
        Err(Error::StepPeerNotFound)
    }
```

File: src/raftstore/store/store.rs [hunk 3]
```diff
@@ -245,19 +245,22 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+        // epoch is stale, we can ignore it.
+        if let Some(peer) = self.region_peers.get(&region_id) {
+            let from_epoch = msg.get_region_epoch();
+            let conf_ver = peer.storage.rl().region.get_region_epoch().get_conf_ver();
+            let version = peer.storage.rl().region.get_region_epoch().get_version();
+
+            if (from_epoch.get_conf_ver() < conf_ver || from_epoch.get_version() < version) &&
+               util::find_peer(&peer.storage.rl().region, from.get_store_id()).is_none() {
+
+                warn!("RequestVote with a stale epoch {:?}, ignore it", from_epoch);
```
Comment (ngaut) [inline src/raftstore/store/store.rs]: How do you know it's a RequestVote message ?

**Comment by siddontang** [comment]
PTAL

**Comment by ngaut** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @BusyJay @disksing

**Comment by BusyJay** [comment]
LGTM

File: src/raftstore/store/store.rs [hunk 4]
```diff
@@ -245,19 +245,24 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
-        if msg.get_message().get_msg_type() == MessageType::MsgRequestVote {
-            if let Some(peer) = self.region_peers.get(&region_id) {
-                let from_epoch = msg.get_region_epoch();
-                // Is it necessary to check version too?
-                if from_epoch.get_conf_ver() <
-                   peer.storage.rl().region.get_region_epoch().get_conf_ver() {
-                    warn!("RequestVote with a stale epoch {:?}, ignore it", from_epoch);
-                    return Ok(());
-                }
+        // If we receive a message which peer is not in current region and
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: s/which/whose sender/

**Pull request merged**